tsl/src/compression/compression.c

```diff
@@ -2239,118 +2236,559 @@ compressed_insert_key_columns(Relation relation)
 	return indexattrs;
 }
 
-void
-decompress_batches_for_insert(const ChunkInsertState *cis, TupleTableSlot *slot)
+/* This method is used to find matching index on compressed chunk
+ * and build scan keys from the slot data
+ */
+static ScanKeyData *
+build_index_scankeys_using_slot(Oid hypertable_relid, Relation in_rel, Relation out_rel,
+								CompressionSettings *settings, TupleTableSlot *slot,
+								Relation *result_index_rel, int *num_scan_keys)
 {
-	/*
-	 * This is supposed to be called with the actual tuple that is being
-	 * inserted, so it cannot be empty.
-	 */
-	Assert(!TTS_EMPTY(slot));
+	List *index_oids;
+	ListCell *lc;
+	ScanKeyData *scankeys = NULL;
+	/* get list of indexes defined on compressed chunk */
+	index_oids = RelationGetIndexList(in_rel);
+	*num_scan_keys = 0;
 
-	Relation out_rel = cis->rel;
+	foreach (lc, index_oids)
+	{
+		Relation index_rel = index_open(lfirst_oid(lc), AccessShareLock);
+		bool matches = false;
 
-	if (!ts_indexing_relation_has_primary_or_unique_index(out_rel))
+		/* Must have at least two attributes. */
+		if (index_rel->rd_index->indnatts < 2)
+		{
+			index_close(index_rel, AccessShareLock);
+			continue;
+		}
+
+		scankeys = palloc0((index_rel->rd_index->indnatts) * sizeof(ScanKeyData));
+
+		for (int i = 0; i < index_rel->rd_index->indnatts; i++)
+		{
+			AttrNumber attnum = AttrOffsetGetAttrNumber(i);
+			char *attname = get_attname(RelationGetRelid(index_rel), attnum, false);
+
+			/* If we are at the last attribute, check its the sequence number attribute.
+			 * This means we found all other attributes on the hypertable and this could be our
+			 * index.
+			 */
+			if (index_rel->rd_index->indnatts - 1 == i)
+			{
+				if (strcmp(attname, COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME) == 0)
+					matches = true;
+				break;
+			}
+
+			if (!ts_array_is_member(settings->fd.segmentby, attname))
+			{
+				break;
+			}
```

Comment:
Let's improve coverage of this. Would be good to test:
* partial index
* covering index
* segmentby columns out of order compared to the table
* sequence number column in the wrong place, e.g. first/middle
* expression index

Reply:
Also a non-btree index.